### PR TITLE
Scripts to automatically package the extensions for Firefox and Chrome.

### PR DESCRIPTION
**Pull request by mossroy**

### CODE REVIEW

File: webextension/backgroundscript.js [hunk 1]
```diff
@@ -36,6 +36,6 @@ genericBrowser.browserAction.onClicked.addListener(handleClick);
 
 function handleClick(event) {
     genericBrowser.tabs.create({
-        url: genericBrowser.runtime.getURL('/index.html')
+        url: genericBrowser.runtime.getURL('/www/index.html')
```
Comment (julianharty) [inline webextension/backgroundscript.js]: I'm curious: is this change a necessary part of packaging the scripts? If not, perhaps it'd be worth separating into a separate commit and pull request?

File: .gitignore [hunk 2]
```diff
@@ -43,3 +43,10 @@ build/config/buildinfo.properties
 
 www-built
 www-ghdeploy
+/node_modules/
+/tmp/
+/scripts/kiwix-*.pem
+/scripts/set_secret_environment_variables.sh
```
Comment (julianharty) [inline .gitignore]: At the risk of stating the obvious, having secret content exposed on an open source repo seems bizarre / counter-productive :)

Comment (julianharty) [inline .gitignore]: I just realised, you protect the secret file using .gitignore - likewise, please ignore my previous comment :)

File: scripts/create_all_packages.sh [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+# So we take the original manifest again, and replace the version inside it again
+cp ../manifest.json ../tmp/
+sed -i -e "s/$VERSION_TO_REPLACE/$VERSION/" ../tmp/manifest.json
+./package_firefox_extension.sh $VERSION
+./package_firefoxos_app.sh $VERSION
+
+CURRENT_DATE=$(date +'%Y-%m-%d')
+# Upload the files on download.kiwix.org
+echo "Uploading the files on http://download.kiwix.org/nightly/$CURRENT_DATE/"
+scp -r -p -i ./travisci_builder_id_key ../build/* nightlybot@download.kiwix.org:/var/www/download.kiwix.org/nightly/$CURRENT_DATE
```
Comment (julianharty) [inline scripts/create_all_packages.sh]: Could anyone use this approach to upload files on download.kiwix.org?

File: scripts/push_to_gh_pages.sh [hunk 4]
```diff
@@ -1 +1,5 @@
+#!/bin/bash
+PREVIOUS_CURRENT_DIR=$(pwd)
+cd ..
 git checkout gh-pages; git merge master; git push; git checkout master
```
Comment (julianharty) [inline scripts/push_to_gh_pages.sh]: what is gh-pages?

File: scripts/create_all_packages.sh (hunk 3 above)
Comment (mossroy) [inline scripts/create_all_packages.sh]: No, because the travisci_builder_id_key (which is necessary for the SSH connection) is not on the repo in clear : it is encrypted with the Travis way to do it

File: scripts/push_to_gh_pages.sh (hunk 4 above)
Comment (mossroy) [inline scripts/push_to_gh_pages.sh]: GitHub can host some web pages, if you simply push them in a gh-pages branch.
I do that on each major release. The result is http://kiwix.github.io/kiwix-html5
It is mainly for testing purpose

File: .gitignore (hunk 2 above)
Comment (mossroy) [inline .gitignore]: Oups, I thought they had been exposed, so I revoked and re-created them.
No matter : it doesn't hurt.

File: webextension/backgroundscript.js (hunk 1 above)
Comment (mossroy) [inline webextension/backgroundscript.js]: It's not 100% necessary, you're right.
It was to avoid adding the /index.html in the packages, which simply redirects to www/index.html.

**Comment by mossroy** [comment]
This fixes #197

File: scripts/package_chrome_extension.sh [hunk 5]
```diff
@@ -0,0 +1,16 @@
...
+VERSION=$1
+cd ..
+echo "Packaging unsigned Chrome extension, version $VERSION"
+zip -r build/kiwix-chrome-unsigned-extension-$VERSION.zip www webextension manifest.json LICENSE-GPLv3.txt service-worker.js README.md
+# Package the extension with Chromium
+# On Travis, we need to make it believe it has a display, because --headless only works on Chromium>=57, which is not available on Travis for now
+if [ ! -z ${TRAVIS_BUILD_DIR+x} ]; then
+    export DISPLAY=:99.0
+    sh -e /etc/init.d/xvfb start
+fi
```
Comment (mgautierfr) [inline scripts/package_chrome_extension.sh]: This could probably be made by travis itself in `.travis.yml`.

Btw, there is a lot of travis specific setting. Setting all of them in `.travis.yml` could be a bit annoying and difficult to maintain. Maybe you could move all travis specific setting in a script `setup_travis_env.sh`

File: scripts/package_firefox_extension.sh [hunk 6]
```diff
@@ -0,0 +1,17 @@
...
+PREVIOUS_CURRENT_DIR=$(pwd)
+VERSION=$1
+cd ../tmp
+echo "Packaging unsigned Firefox extension, version $VERSION"
+zip -r ../build/kiwix-firefox-unsigned-extension-$VERSION.xpi www webextension manifest.json LICENSE-GPLv3.txt service-worker.js README.md
+# Install web-ext if it's not already installed
+if [ ! -f ../node_modules/web-ext/bin/web-ext ]; then
+    cd ..
+    npm install web-ext
+    cd tmp
```
Comment (mgautierfr) [inline scripts/package_firefox_extension.sh]: All those `cd` are difficult to follow.

If you assume that the script is run in the right directory you could write this part like this :
```
REVIOUS_CURRENT_DIR=$(pwd)
VERSION=$1

# Install web-ext if it's not already installed
if [ ! -f node_modules/web-ext/bin/web-ext ]; then
    npm install web-ext
endif

cd tmp
echo "Packaging unsigned Firefox extension, version $VERSION"
zip -r ../build/kiwix-firefox-unsigned-extension-$VERSION.xpi www webextension manifest.json LICENSE-GPLv3.txt service-worker.js README.md
[...]
```

File: webextension/backgroundscript.js (hunk 1 above)
Comment (mgautierfr) [inline webextension/backgroundscript.js]: > It was to avoid adding the /index.html in the packages, which simply redirects to www/index.html.

It would have been nice to have this explanation in the commit message :)

**Comment by mgautierfr** [comment]
> Add application id for Firefox (even if it annoys Chrome) 

How does it annoy Chrome ?

**Comment by mossroy** [comment]
I've pushed fixes regarding your comments, except for the mktemp because it would force to pass this directory as a parameter to all the scripts

**Comment by mgautierfr** [comment]
Looks good to me. 
Nice work !

**Comment by mossroy** [comment]
Thanks a lot for the review and the hints
